pkg/roles/service.go

```diff
@@ -0,0 +1,310 @@
+	ctx, span := s.tracer.Start(ctx, "roles.Service.RemovePermissions")
+	defer span.End()
+
+	// preemptive check to verify if all permissions to be assigned are accessible by the user
+	// needs to happen separately
+
+	ps := make([]ofga.Tuple, 0)
+
+	for _, p := range permissions {
+		ps = append(ps, *ofga.NewTuple(fmt.Sprintf("role:%s#%s", ID, ASSIGNEE_RELATION), p.Relation, p.Object))
+	}
+
+	err := s.ofga.DeleteTuples(ctx, ps...)
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) DeleteRole(ctx context.Context, ID string) error {
+	ctx, span := s.tracer.Start(ctx, "roles.Service.DeleteRole")
+	defer span.End()
+
+	permTypes := []string{"role", "group", "identity", "scheme", "provider", "client"}
+	var wg sync.WaitGroup
+
+	wg.Add(len(permTypes))
+
+	// TODO @shipperizer use a background operator
+	for _, t := range permTypes {
+		go func(pType string) {
+			defer wg.Done()
+			s.removePermissionsByType(ctx, ID, pType)
+		}(t)
+	}
+
+	wg.Wait()
+
+	return s.ofga.DeleteTuples(ctx, *ofga.NewTuple(authorization.ADMIN_PRIVILEGE, "privileged", fmt.Sprintf("role:%s", ID)))
+}
+
+// TODO @shipperizer make this more scalable by pushing to a channel and using goroutine pool
+// potentially create a background operator that can pipe results to an on demand channel and works off a
+// set amount of goroutines
+func (s *Service) listPermissionsByType(ctx context.Context, roleIDAssignee, pType, continuationToken string) ([]string, string, error) {
+	ctx, span := s.tracer.Start(ctx, "roles.Service.listPermissionsByType")
+	defer span.End()
+
+	r, err := s.ofga.ReadTuples(ctx, roleIDAssignee, "", fmt.Sprintf("%s:", pType), continuationToken)
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, "", err
+	}
+
+	permissions := make([]string, 0)
+
+	for _, t := range r.GetTuples() {
+		permissions = append(permissions, authorization.NewUrn(t.Key.Relation, t.Key.Object).ID())
+	}
+
+	return permissions, r.GetContinuationToken(), nil
+}
+
+func (s *Service) removePermissionsByType(ctx context.Context, ID, pType string) {
+	ctx, span := s.tracer.Start(ctx, "roles.Service.removePermissionsByType")
+	defer span.End()
+
+	cToken := ""
+	assigneeRelation := fmt.Sprintf("role:%s#%s", ID, ASSIGNEE_RELATION)
+	permissions := make([]ofga.Tuple, 0)
+	for {
+		r, err := s.ofga.ReadTuples(ctx, assigneeRelation, "", fmt.Sprintf("%s:", pType), cToken)
+
+		if err != nil {
+			s.logger.Errorf("error when retrieving tuples for %s %s", assigneeRelation, pType)
+			return
+		}
+
+		for _, t := range r.Tuples {
+			permissions = append(permissions, *ofga.NewTuple(assigneeRelation, t.Key.Relation, t.Key.Object))
+		}
+
+		// if there are more pages, keep going with the loop
+		if cToken = r.ContinuationToken; cToken != "" {
+			continue
+		}
+
+		// TODO @shipperizer understand if better breaking at every cycle or reverting if clause
+		break
+	}
+
+	if err := s.ofga.DeleteTuples(ctx, permissions...); err != nil {
```

Comment:
Perhaps it is an unrealistic case, but I am a little worried that this may error if too many permissions are returned. I mean that openfga chooses not to return the permissions all at once, it is possible that it will not be able to handle them all at once either. In the API docs there is no mention of a limit, but I think that we should test it. 

Disregard this if it does not make sense or leave a TODO to look into it in the future. 

If this could cause an issue we could create a go routine for deleting the permissions for every continuation token we get, eg:
```
for {
  // Get the permissions using the continuation token
  permissions, error := getPermissions(continuationToken)
  // Delete permissions without blocking
  go deletePermissions(permissions)
  cToken = permissions.ContinuationToken
  if cToken == "" {
    break
  }
  // wait for all routines to return
}
```

Reply:
issue is we have already enough loose goroutines per request (6) so this would be really unmanageable with the current implementation

we need to find a better setup to deal with the request load, that would be the first step

ideally we would create a `TaskManager` object that deals with a pool of goroutines and allows for this background work to happen, but i can see it quite complicated 

i'll definitely add a TODO

Reply:
https://github.com/canonical/identity-platform-admin-ui/issues/217